### PR TITLE
Openshift Mode Annotation

### PR DESCRIPTION
**Pull request by waveywaves**
- Added annotation to work with Openshift `jenkins.dev/openshift-mode`
- Adding `"jenkins.io/openshift-mode": "true"` allows the openshift-jenkins image to run smoothly on the Operator with the appropriate `JENKINS_HOME` env
- Annotation disables checks for imagepullsecret, security context, pod security policy, and command

### CODE REVIEW

**Comment by waveywaves** [comment]
cc @tomaszsek PTAL

File: pkg/controller/jenkins/configuration/base/reconcile.go [hunk 1]
```diff
@@ -853,3 +854,14 @@ func (r *ReconcileJenkinsBaseConfiguration) ensureBaseConfiguration(jenkinsClien
 
 	return reconcile.Result{Requeue: requeue}, err
 }
+
+func (r *ReconcileJenkinsBaseConfiguration) isOpenshiftEnabled() bool {
+	openshiftEnabled := false
+	if r.Configuration.Jenkins.Annotations[openshiftEnabledAnnotation] == "true" {
+		openshiftEnabled = true
+		r.logger.Info("-Openshift Mode-")
```
Comment (jakalkhalili) [inline pkg/controller/jenkins/configuration/base/reconcile.go]: I don't think that printing to the log is necessary. Also every time you call `isOpenshiftEnabled` function this line will print the message every time that function is being called.

File: pkg/controller/jenkins/configuration/base/reconcile.go [hunk 2]
```diff
@@ -853,3 +854,14 @@ func (r *ReconcileJenkinsBaseConfiguration) ensureBaseConfiguration(jenkinsClien
...
+	} else {
+		openshiftEnabled = false
+	}
```
Comment (jakalkhalili) [inline pkg/controller/jenkins/configuration/base/reconcile.go]: remove this, this doesn't do anything. When removed, if the condition on line 860 won't be satisfied, then the `openshiftEnabled` will be `false`

File: pkg/controller/jenkins/configuration/base/reconcile.go [hunk 3]
```diff
@@ -853,3 +854,14 @@ func (r *ReconcileJenkinsBaseConfiguration) ensureBaseConfiguration(jenkinsClien
...
+	return openshiftEnabled
```
Comment (jakalkhalili) [inline pkg/controller/jenkins/configuration/base/reconcile.go]: I think this function would be shortened to

```
func (r *ReconcileJenkinsBaseConfiguration) isOpenshiftEnabled() bool {
    return bool(r.Configuration.Jenkins.Annotations[openshiftEnabledAnnotation])
}
```

Or something like this

File: pkg/controller/jenkins/configuration/base/reconcile.go [hunk 4]
```diff
@@ -609,23 +610,23 @@ func (r *ReconcileJenkinsBaseConfiguration) compareContainers(expected corev1.Co
...
 	}
 	if !reflect.DeepEqual(expected.Resources, actual.Resources) {
 		messages = append(messages, "Resources have changed")
 		verbose = append(verbose, fmt.Sprintf("Resources have changed to '%+v' in container '%s'", expected.Resources, expected.Name))
 	}
-	if !reflect.DeepEqual(expected.SecurityContext, actual.SecurityContext) {
+	if !reflect.DeepEqual(expected.SecurityContext, actual.SecurityContext) && !r.isOpenshiftEnabled() {
 		messages = append(messages, "Security context has changed")
 		verbose = append(verbose, fmt.Sprintf("Security context has changed to '%+v' in container '%s'", expected.SecurityContext, expected.Name))
 	}
```
Comment (jakalkhalili) [inline pkg/controller/jenkins/configuration/base/reconcile.go]: Add `!r.isOpenshiftEnabled()` at beginning and then check the other conditions.
Something like this:

```
if !r.isOpenshiftEnabled() {
    if !reflect.DeepEqual(expected.LivenessProbe, actual.LivenessProbe)...
    if !reflect.DeepEqual(expected.ReadinessProbe, actual.ReadinessProbe)...
    if !reflect.DeepEqual(expected.SecurityContext, actual.SecurityContext)...

   [...]
}

```

File: pkg/controller/jenkins/configuration/base/reconcile.go [hunk 5]
```diff
@@ -36,6 +36,7 @@ import (
 
 const (
 	fetchAllPlugins = 1
+	openshiftEnabledAnnotation = "jenkins.dev/openshift-mode"
```
Comment (jakalkhalili) [inline pkg/controller/jenkins/configuration/base/reconcile.go]: `jenkins.dev/openshift-mode`, I think this is not the good name of annotation. Something like `jenkins.jenkins.io/openshift-enabled`. @tomaszsek @pawelprazak what do you think?

File: pkg/controller/jenkins/configuration/base/reconcile.go (hunk 1 above)
Comment (waveywaves) [reply pkg/controller/jenkins/configuration/base/reconcile.go]: makes sense I'll do that

File: pkg/controller/jenkins/configuration/base/reconcile.go (hunk 2 above)
Comment (waveywaves) [reply pkg/controller/jenkins/configuration/base/reconcile.go]: good catch on it

File: pkg/controller/jenkins/configuration/base/reconcile.go (hunk 4 above)
Comment (waveywaves) [reply pkg/controller/jenkins/configuration/base/reconcile.go]: This doesn't make sense, as I suppose while checking this stuff you guys must ahve kept the order in mind, for what resources are being checked, if not I can do it this way.

File: pkg/controller/jenkins/configuration/base/reconcile.go (hunk 5 above)
Comment (waveywaves) [reply pkg/controller/jenkins/configuration/base/reconcile.go]: How about just `jenkins.io/openshift-enabled` ?

File: pkg/controller/jenkins/configuration/base/reconcile.go (hunk 3 above)
Comment (waveywaves) [reply pkg/controller/jenkins/configuration/base/reconcile.go]: let's not typecast, not a good practice.

File: pkg/controller/jenkins/configuration/base/reconcile.go (hunk 1 above)
Comment (waveywaves) [reply pkg/controller/jenkins/configuration/base/reconcile.go]: come to think of it I will change the var to `isOpenshift` and keep `openshift mode` as openshift is not something that is enabled if that makes sense

**Comment by tomaszsek** [comment]
Like we discussed, lets try first general approach (#190), without OpenShift specific modifications, and if that fails will think of an architecture that would allow for custom providers.

In the meantime you could pick #194 or #199 if you'd like.

**Comment by waveywaves** [comment]
@tomaszsek Okay, I will open a PR accordingly

**Pull request closed without merging**